client.go

```diff
@@ -156,13 +164,16 @@ func (c *Client) Push(n *Notification) (*Response, error) {
 // return a Response indicating whether the notification was accepted or
 // rejected by the APNs gateway, or an error if something goes wrong.
 func (c *Client) PushWithContext(ctx Context, n *Notification) (*Response, error) {
-	payload, err := json.Marshal(n)
-	if err != nil {
+	payload := payloads.Get().(*bytes.Buffer)
+	payload.Reset()
+	defer payloads.Put(payload)
+	if err := json.NewEncoder(payload).Encode(n); err != nil {
 		return nil, err
 	}
+	payload.Truncate(payload.Len() - len("\n"))
 
 	url := fmt.Sprintf("%v/3/device/%v", c.Host, n.DeviceToken)
-	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(payload))
+	req, err := http.NewRequest(http.MethodPost, url, payload)
```

Comment:
Thanks for this PR. i like what this is trying to do as it must be expensive doing the encoding every time. Wondering if an alternative is to just have an extra method that is `PushBytesWithContext()` or similar. That means that we don't have to do json encoding inside the `PushWithContext` and you could handle that outside the method. This would reduce allocations and give you the flexibility to do what you're doing but still keep the code simple.

Reply:
Actually that makes no sense, as payload is nested

Reply:
But maybe it's over-engineering

Reply:
actully we can do just MarshalJSON on notirifcation

Reply:
Pooling bytes makes more sense in case of https://github.com/sideshow/apns2/pull/206/files